### PR TITLE
Reuse neighborqueue during hnsw index build (attempt 2)

### PR DESCRIPTION
**Pull request by jbellis**
This changes HnswGraphBuilder to re-use the same candidates queues for adding nodes by allocating them in the Builder instance.

This saves about 2.5% of build time and takes memory allocations of NQ long[] from 25% of total to 0%.  JFR runs are attached.

The difference from the first attempt (which actually made things slower for some graphs) is that it preserves the original code's behavior of using a 1-sized queue for the search in the levels above where the node actually gets added.

[main.jfr.gz](https://github.com/apache/lucene/files/11749837/main.jfr.gz)
[nq2.jfr.gz](https://github.com/apache/lucene/files/11749838/nq2.jfr.gz)

### CODE REVIEW

**Comment by jbellis** [comment]
Additionally, the original change only re-used the candidates queues within a single addNode call, so this is improved in that respect as well.

**Comment by jbellis** [comment]
cc @msokolov @benwtrent @zhaih

**Comment by benwtrent** [comment]
Hey @jbellis the change looks nice to me. But, I ran https://github.com/mikemccand/luceneutil `knnPerTest` and saw no change at all in indexing time.

Am I missing something? Could you provide the benchmark you ran to track the 2.5% improvement?

**Comment by jbellis** [comment]
I'm using the million-row sift dataset via this harness https://github.com/jbellis/hnswdemo/tree/benchmarking

I believe what is happening is that allocation is basically free and there is enough slack across the JVM in these synthetic benchmarks to soak up the extra GC required -- I do not see any changes when run normally, either.  So I forced it to a single core with

```$ taskset -c 0-1 ./gradlew runTexmex -PsiftName=sift```

(it is a hyperthreaded core so i give it logical cores 0-1)

Also if you take a look at the jfr files it is very clear that a significant amount of allocation is gone.

File: lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java [hunk 1]
```diff
@@ -262,7 +274,6 @@ private NeighborQueue searchLevel(
       int visitedLimit)
       throws IOException {
     int size = graph.size();
-    NeighborQueue results = new NeighborQueue(topK, false);
```
Comment (benwtrent) [inline lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphSearcher.java:265]: A minor thing, could we `assert results.isMinHeap()` or something? Of course `isMinHeap()` would be a package private method.

Just to catch bad usage within tests if somebody accidentally uses a max heap here.

**Review by benwtrent — APPROVED** [review]
Change seems sane to me. Thanks for the optimizations!

**Comment by jbellis** [comment]
CI says this test is failing

```org.apache.lucene.backward_index.TestBackwardsCompatibility > test suite's output saved to /home/runner/work/lucene/lucene/lucene/backward-codecs/build/test-results/test/outputs/OUTPUT-org.apache.lucene.backward_index.TestBackwardsCompatibility.txt, copied below:
   >     java.lang.AssertionError: Missing backcompat test files:
   >       9.6.0-cfs
   >       9.7.0-cfs
   >         at __randomizedtesting.SeedInfo.seed([6833F9FB78703710:78EC16BBCCE7213C]:0)
   >         at junit@4.13.1/org.junit.Assert.fail(Assert.java:89)
   >         at org.apache.lucene.backward_index.TestBackwardsCompatibility.testAllVersionsTested(TestBackwardsCompatibility.java:818)
```

Not sure how that's related to this code or how to fix it

tests pass locally

**Comment by zhaih** [comment]
> Not sure how that's related to this code or how to fix it

I think this is due to newly cut 9.7 branch so we probably just need to wait a bit more. I see all the auto testing is complaining too.

**Review by zhaih — APPROVED** [review]
Thank you for pursuing this! LGTM

**Comment by benwtrent** [comment]
@jbellis if you are ready, I can merge and backport this change.

**Pull request merged**